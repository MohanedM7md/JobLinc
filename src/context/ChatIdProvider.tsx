import React, { createContext, useContext, useState, useEffect } from "react";

interface ChatIdContextType {
  chatId: string;
  setChatId: React.Dispatch<React.SetStateAction<string>>;
}
export const ChatsIdContext = createContext<ChatIdContextType>({
  chatId: "",
  setChatId: () => {},
});

export function ChatIdProvider({
  children,
  id,
}: {
  children: React.ReactNode;
  id?: string;
}) {
  const [chatId, setChatId] = useState(id ? id : "");

  return (
    <ChatsIdContext.Provider value={{ chatId, setChatId }}>
      {children}
    </ChatsIdContext.Provider>
  );
}

export default function useChatId() {
  const context = useContext(ChatsIdContext);
  if (!context) throw new Error("useChats must be used within a ChatProvider");

  return context;
}
