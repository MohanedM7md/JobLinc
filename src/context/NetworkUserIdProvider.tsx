import React, { createContext, useContext, useState } from "react";

interface ChatIdContextType {
  title: string;
  setTitle: React.Dispatch<React.SetStateAction<string>>;
  usersId: string[];
  setUsersId: React.Dispatch<React.SetStateAction<string[]>>;
}

export const ChatsIdContext = createContext<ChatIdContextType>({
  title: "",
  setTitle: () => {},
  usersId: [],
  setUsersId: () => {},
});

interface NetworkUserIdProviderProps {
  children: React.ReactNode;
  ids?: string[];
  initialTitle?: string;
}

export function NetworkUserIdProvider({
  children,
  ids,
  initialTitle = "",
}: NetworkUserIdProviderProps) {
  const [usersId, setUsersId] = useState<string[]>(ids ?? []);
  const [title, setTitle] = useState<string>(initialTitle);

  return (
    <ChatsIdContext.Provider value={{ title, setTitle, usersId, setUsersId }}>
      {children}
    </ChatsIdContext.Provider>
  );
}

export default function useNetworkUserId() {
  const context = useContext(ChatsIdContext);
  if (!context)
    throw new Error(
      "useNetworkUserId must be used within a NetworkUserIdProvider",
    );

  return context;
}
