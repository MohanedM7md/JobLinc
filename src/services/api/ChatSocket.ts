import { connectSocket } from "./socket";
import { RecievedMessage } from "@chatComponent/interfaces/Message.interfaces";
import { ChatCardInterface } from "@chatComponent/interfaces/Chat.interfaces";
import toast from "react-hot-toast";

let ChatSocket: SocketIOClient.Socket | null = null;

export const connectToChat = () => {
  ChatSocket = connectSocket("chat");
  return ChatSocket;
};
export const onConnect = (setTrue: (boolean: boolean) => void) => {
  if (!ChatSocket) return;
  ChatSocket?.on("connect", () => {
    setTrue(true);
    ChatSocket?.on("error", (error: { event: string; message: string }) => {
      console.log(error.message);
      toast.error(error.message);
    });
    ChatSocket?.on("notify", (error: { event: string; message: string }) => {
      console.log(error.message);
      toast.error(error.message);
    });
  });
};
export const subscribeToMessages = (
  chatId: string,
  onMessageReceived: (message: RecievedMessage) => void,
  onMessageRead: (senderId: string) => void,
  onUserTyping: (userId: string) => void,
  onStopUserTyping: (userId: string) => void,
) => {
  if (!ChatSocket) return;

  ChatSocket.emit("openChat", chatId);
  ChatSocket.on("receiveMessage", (message: RecievedMessage) => {
    console.log("ðŸ“© Received Message:", message);
    ChatSocket!.emit("messageReceived", chatId, message.messageId);
    onMessageReceived(message);
  });

  ChatSocket.on("messageRead", (senderId: string) => {
    console.log("ðŸ“– Message Read by", senderId);
    onMessageRead(senderId);
  });
  ChatSocket.on("messageTyping", (userId: string) => {
    console.log("ðŸ“– Other User is typing :", userId);
    onUserTyping(userId);
  });
  ChatSocket.on("stopTyping", (userId: string) => {
    console.log("ðŸ“– Other User is stopped typing : ", userId);
    onStopUserTyping(userId);
  });
};

export const sendMessage = (
  chatId: string,
  message: RecievedMessage,
  callback?: () => void,
) => {
  if (!ChatSocket) return;

  ChatSocket.emit("sendMessage", { ...message, chatId }, callback);
  console.log("ðŸ“© Sent Message:", message);
};
export const typing = (chatId: string) => {
  if (!ChatSocket) return;
  ChatSocket.emit("messageTyping", chatId);
  console.log("ðŸ“– I am Typing in chat:", chatId);
};
export const stopTyping = (chatId: string) => {
  if (!ChatSocket) return;
  ChatSocket.emit("stopTyping", chatId);
  console.log("ðŸ“– I stopped Typing ");
};
export const unsubscribeFromMessages = (chatId: string) => {
  if (!ChatSocket) return;
  ChatSocket.off("receiveMessage");
  ChatSocket.off("messageRead");
  ChatSocket.emit("leaveChat", chatId);
};

export const disconnectChatSocket = () => {
  if (ChatSocket) {
    ChatSocket.off("error");
    ChatSocket.off("notify");
    ChatSocket.disconnect();
  }
};
export const subscribeToChats = (
  onChatUpdate: (chatCard: ChatCardInterface) => void,
  onNewChat: (chatCard: ChatCardInterface) => void,
) => {
  if (ChatSocket) console.log("i am listing for any Chat card changes");
  else console.log("El socket chat card listing is died");
  ChatSocket?.on("cardUpdate", (chatCard: ChatCardInterface) => {
    console.log("ðŸ“© Received modified chatCard:", chatCard);
    onChatUpdate(chatCard);
  });
  ChatSocket?.on("newChat", (chatCard: ChatCardInterface) => {
    console.log("ðŸ“© Received new chatCard:", chatCard);
    onNewChat(chatCard);
  });
};
export const listenToOpenChatErrors = () => {};
export default connectToChat;
