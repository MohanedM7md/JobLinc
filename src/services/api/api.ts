import axios from "axios";
import store from "@store/store";
import { API_URL } from "./config";
import { logOut, updateAccessToken } from "@store/user/userSlice";

export const api = axios.create({
  baseURL: API_URL,
  headers: {
    "Content-Type": "application/json",
  },
});

api.interceptors.request.use(
  (config) => {
    const token = store.getState().user.accessToken;
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error),
);

api.interceptors.response.use(
  (response) => response,
  async (error) => {
    const originalRequest = error.config;
    if (
      error.response &&
      error.response.status === 401 &&
      (error.response.data.errorCode === 401100 ||
        error.response.data.errorCode === 401101) &&
      !originalRequest._retry
    ) {
      originalRequest._retry = true;

      const dispatch = store.dispatch;
      try {
        console.log("refreshing refresh token");
        const refreshToken = localStorage.getItem("refreshToken");
        const userId = localStorage.getItem("userId");
        const companyId = localStorage.getItem("companyId");

        if (!refreshToken && !userId) {
          console.log("No refresh token found, logging out...");
          localStorage.removeItem("refreshToken");
          localStorage.removeItem("userId");
          dispatch(logOut());

          return Promise.reject(error);
        }
        let data;
        try {
          let response;
          if (companyId)
            response = await api.post("auth/refresh-token", {
              userId,
              refreshToken,
              companyId,
            });
          else
            response = await api.post("auth/refresh-token", {
              userId,
              refreshToken,
            });
          data = response.data;
        } catch (refreshError) {
          console.log("Failed to refresh token, logging out...");
          dispatch(logOut());
          window.location.href = "/signin";
          return Promise.reject(refreshError);
        }
        dispatch(updateAccessToken(data.accessToken));
        localStorage.setItem("refreshToken", data.refreshToken);
        originalRequest.headers.Authorization = `Bearer ${data.accessToken}`;
        return api(originalRequest);
      } catch (refreshError) {
        console.log("Refresh token expired, logging out...");
        dispatch(logOut());
        window.location.href = "/signin";
        return Promise.reject(refreshError);
      }
    }
    return Promise.reject(error);
  },
);
