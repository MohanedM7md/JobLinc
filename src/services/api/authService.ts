import { api } from "@services/api/api";
import store from "@store/store";
import { updateAccessToken } from "@store/user/userSlice";

export const loginUserAPI = async (userData: {
  email: string;
  password: string;
}) => {
  try {
    const response = await api.post("auth/login", userData);
    return response.data;
  } catch (err) {
    console.log(err);
  }
};

export const loginWithGoogleAPI = async(userCredentials: {credential: string}) => {
  try {
    const response = await api.post("auth/google-login", userCredentials);
    return response.data;
  }
  catch(err)
  {
    console.error(err);
  }
}

export const registerUserAPI = async (userData: {
  firstname: string;
  lastname: string;
  email: string;
  password: string;
  country: string;
  city: string;
  phoneNumber: string;
}) => {
  const response = await api.post("auth/register", userData);
  return response.data;
};

export const forgotPasswordAPI = async (userData: { email: string }) => {
  const response = await api.post("auth/forgot-password", userData);
  return response.data;
};

export const resetPasswordAPI = async (userData: {
  email: string;
  newPassword: string;
  resetToken: string;
}) => {
  const response = await api.post("auth/reset-password", userData);
  return response.data;
};

export const confirmOTPAPI = async (userData: {
  email: string;
  forgotToken: string;
  otp: string;
}) => {
  const response = await api.post("auth/confirm-otp", userData);
  return response.data;
};

export const changePasswordAPI = async (userData: {
  oldPassword: string;
  newPassword: string;
  refreshToken: string;
}) => {
  const response = await api.post("auth/change-password", userData);
  return response.data;
};

export const updateEmailAPI = async (userData: {
  userId: string;
  email: string;
}) => {
  const response = await api.put("user/edit/email", userData);
  return response.data;
};

export const updateUsernameAPI = async (userData: {
  userId: string;
  username: string;
}) => {
  const response = await api.put("user/edit/username", userData);
  return response.data;
};

export const getUserDetailsAPI = async () => {
  const response = await api.get("user/me");
  console.log("response of user Details yarb: ", response.data);
  return response.data;
};

export const sendConfirmationEmailAPI = async (userData: { email: string }) => {
  const response = await api.post("auth/send-confirmation-email", userData);
  return response.data;
};

export const confirmEmailAPI = async (userData: {
  email: string;
  token: string;
  otp: string;
}) => {
  const response = await api.post("auth/confirm-email", userData);
  return response.data;
};

export const refreshToken = async () => {
  const refreshToken = localStorage.getItem("refreshToken");
  const userId = localStorage.getItem("userId");
  const response = await api.post("auth/refresh-token", {
    userId,
    refreshToken,
  });
  if (response.status !== 200)
  {
    window.location.href = "/home";
  }
  const dispatch = store.dispatch;
  const { data } = response;
  dispatch(updateAccessToken(data.accessToken));
  localStorage.setItem("refreshToken", data.refreshToken);
}
