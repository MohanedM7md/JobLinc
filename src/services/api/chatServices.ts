import { api } from "./api";
import axios from "axios";
import toast from "react-hot-toast";
export const fetchChats = async () => {
  const response = await api.get("/chat/all");
  return response.data;
};

export const fetchNetWorks = async (Id: string) => {
  const response = await api.get(`/connection/connected`);
  const users = response.data.map(
    ({
      userId,
      firstname,
      lastname,
      profilePicture: chatPicture,
    }: {
      userId: string;
      firstname: string;
      lastname: string;
      profilePicture: string;
    }) => ({ userId, chatName: `${firstname} ${lastname}`, chatPicture }),
  );
  return users;
};
export const fetchChatData = async (chatId: string) => {
  const response = await api.get(`/chat/c/${chatId}`);
  return response.data;
};
export const createChat = async (
  recievers: string[],
  title: string | undefined,
) => {
  try {
    const response = await api.post(`/chat/create`, {
      receiverIds: recievers,
      title,
    });
    toast.success("Chat succefully created!");
    return response;
  } catch (error) {
    if (axios.isAxiosError(error)) {
      if (error.response?.status === 429) {
        toast.error("Too many requests! Please upgrade your plan.");
        throw "429";
      } else {
        const errorMessage =
          error.response?.data?.message || error.message || "Unknown error";
        toast.error(`Error: ${errorMessage}`);
        throw new Error(errorMessage);
      }
    } else {
      const errorMessage =
        error instanceof Error ? error.message : "Unknown error";
      toast.error(`Error: ${errorMessage}`);
      throw error;
    }
  }
};

export const ReadToggler = async (chatId: string) => {
  const response = await api.put(`/chat/readOrUnread`, { chatId });
  return response.data;
};

export const uploadingMedia = async (file: File): Promise<string> => {
  const formData = new FormData();
  formData.append("file", file);

  try {
    const response = await api.post(`/chat/upload-media`, formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    });
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error) && error.response?.status === 400) {
      toast.error("File not supported");
      throw new Error("File not supported");
    } else {
      const errorMessage =
        error instanceof Error ? error.message : "Unknown error";
      toast.error(`Error: ${errorMessage}`);
      throw error;
    }
  }
};

export const fetchRequestChatData = async () => {
  const response = await api.get("/chat/message-requests");
  if (response.status != 200)
    throw new Error("Failed to fetch message request chat data");
  return response.data;
};

export const chatRequestStatus = async (chatId: string, status: string) => {
  const response = await api.put("/chat/change-request-status", {
    chatId,
    status,
  });
  if (response.status != 200)
    throw new Error("Failed to change message status");
  return response.data;
};

export const BlockMessaging = async (status: boolean) => {
  const response = await api.put("/user/edit/personal-info", {
    allowMessages: status,
  });
  if (response.status != 200)
    throw new Error(`Response error:${response.status} failed to block use`);
};

export const getConnections = async (chatId: string) => {
  const response = await api.get(`/chat/available-connections/${chatId}`);
  if (!(response.status == 200))
    throw new Error(
      `error status:${response.status} Error while fetching connections`,
    );
  return response.data;
};

export const addParticipants = async (
  participants: string[],
  chatId: string,
) => {
  const response = await api.post("/chat/add-or-remove-participants", {
    userIds: participants,
    chatId,
    action: "add",
  });
  if (response.status != 200) {
    toast.error(`sdasdasda${response.statusText}`);
    throw new Error(
      `Response error: ${response.status} - Failed to remove participants`,
    );
  }
  toast.success("paricipant successfuly added!");
  return response.data;
};

export const removeParticipants = async (
  participants: string[],
  chatId: string,
) => {
  const response = await api.post("/chat/add-or-remove-participants", {
    userIds: participants,
    chatId,
    action: "remove",
  });
  if (response.status != 200) {
    toast.error(`${response.statusText}`);
    throw new Error(
      `Response error: ${response.status} - Failed to remove participants`,
    );
  }
  toast.success("paricipant successfuly added!");
  return response.data;
};
