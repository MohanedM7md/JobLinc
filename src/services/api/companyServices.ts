import { api } from "./api";
import { FormData_Interface } from "../../components/Company/interfaces/inputs.interface";
import store from "@store/store";

import { updateAccessToken } from "@store/user/userSlice";
export const slugChecker = async (slug: string) => {
  const response = await api.post(`/companies/${slug}`, {});
  return response.data;
};
export const enterAdminPage = async (companyId: string) => {
  const userId = localStorage.getItem("userId");
  const refreshToken = localStorage.getItem("refreshToken");
  const response = await api.post("auth/refresh-token", {
    userId,
    companyId,
    refreshToken,
  });
  if (response.status != 200)
    throw new Error("Failed to refresh token. Please try again.");
  const { data } = response;
  store.dispatch(updateAccessToken(data.accessToken));
  localStorage.setItem("refreshToken", data.refreshToken);
  localStorage.setItem("companyId", companyId);
};
export const createCompnay = async (formData: FormData_Interface) => {
  console.log("Form Data", formData);
  const response = await api.post(`/companies`, { ...formData });
  return response;
};

export const updateInfo = async (formData: FormData_Interface) => {
  console.log("Form Data", formData);
  const response = await api.patch(`/companies`, { ...formData });
  return response;
};

export const checkUniqueSlug = async (slug: string) => {
  const response = await api.get(`/companies/check-unique-slug/${slug}`);
  return response;
};

export const getMyCompany = async () => {
  const response = await api.get(`/companies/me`);
  return response;
};

export const getAllCompanies = async () => {
  const response = await api.get('/companies');
  return response;
}

export const getMyCompanyFollowers = async () => {
  const response = await api.get(`/follow/followers`);
  return response;
};

export const getCompanyBySlug = async (slug: string) => {
  const response = await api.get(`/companies/${slug}`);
  return response;
};

export const uploadFile = async (file: File): Promise<string> => {
  const formData = new FormData();
  formData.append("file", file);

  const response = await api.post(`/chat/upload-media`, formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  });

  return response.data;
};

export const addAdmin = async (userId: string) => {
  const response = await api.post(`/companies/admins`, { userId: userId });
  return response;
};

export const deleteAdmin = async (userId: string) => {
  const response = await api.delete(`/companies/admins/${userId}`);
  return response;
};

export const getConnections = async () => {
  const response = await api.get("/connection/connected");
  if (response.status != 200)
    throw new Error("Failed to fetch connection. Please try again.");
  return response.data;
};

export async function searchCompanies({name, values}:{name?: string, values?: string}) {
  try {
    let params;
    params = name ? { search: name } : {};
    params = values ? {...params, fields: values} : {...params}
    const response = await api.get(`companies`, { params });
    return response.data;
  } catch (error) {
    console.log("Error fetching companies:", error);
    throw error;
  }
}

export async function deleteCompany()
{
  try {
    const response = await api.delete("/companies");
    return response;
  }
  catch(error)
  {
    console.error("Error cannot delete company ", error);
  }
}
