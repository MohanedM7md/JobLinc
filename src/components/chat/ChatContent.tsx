import { memo, useState, useEffect, useRef } from "react";
import ChatMessages from "./ChatMessages";
import ChatInput from "./ChatInput";
import {
  subscribeToMessages,
  unsubscribeFromMessages,
  sendMessage,
  typing,
  stopTyping,
  listenToOpenChatErrors,
} from "@services/api/ChatSocket";
import { RecievedMessage } from "./interfaces/Message.interfaces";
import { User } from "./interfaces/User.interfaces";
import useChats from "@hooks/useChats";
import { fetchChatData, createChat } from "@services/api/chatServices";
import useChatid from "@context/ChatIdProvider";
import useNetworkUserId from "@context/NetworkUserIdProvider";
import UserTypingIndicator from "./UserTyping";
import GroupChatSetting from "./GroupSetting/GroupChatSetting";
import { toast } from "react-toastify"; // Ensure you're importing toast
import ChatError from "./ChatError";
function ChatContent({
  className,
  onClose,
}: {
  className?: string;
  onClose?: () => void;
}) {
  const [users, setUsers] = useState<User[]>([]);
  const [messages, setMessages] = useState<RecievedMessage[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  const { chatId, setChatId } = useChatid();
  const { usersId } = useNetworkUserId();
  const { title } = useNetworkUserId();

  const userIdRef = useRef<string | null>(localStorage.getItem("userId"));
  const messagesEndRef = useRef<HTMLDivElement | null>(null);
  const [typingUsers, setTypingUsers] = useState<string[]>([]);
  const chatType = useRef<string>("");
  let messageDelivered = useRef<boolean>(false);

  const handleMessageStatus = () => {
    setTimeout(() => {
      if (!messageDelivered.current) {
        messageDelivered.current = true;
        setMessages((prevMsgs) =>
          prevMsgs.map((msg) =>
            msg.messageId === messageId ? { ...msg, status: "failed" } : msg,
          ),
        );
      }
    }, 1000);
  };

  useEffect(() => {
    messagesEndRef.current?.scrollIntoView({ behavior: "smooth" });
  }, [messages]);

  useEffect(() => {
    const fetchData = async () => {
      setLoading(true);
      setError(null);
      try {
        if (usersId.length && !chatId) {
          console.log("Create new chat with: ", usersId);
          try {
            const { data } = await createChat(usersId, title);
            setUsers(data.participants);
            console.log(data.participants);
            setMessages(data.messages);
            setChatId(data.chatId);

            chatType.current = data.chatType;
          } catch (err) {
            console.error("Failed to create chat:", err);
            if (err == "429")
              setError(" انت الان على باقة الفقراء مسموحلك 5 شاتس فاليوم بس");
            else setError("Failed to create chat");
            onClose?.();
            return;
          }
        } else if (chatId) {
          try {
            const data = await fetchChatData(chatId);
            setUsers(data.participants);
            setMessages(data.messages);
            chatType.current = data.chatType;
          } catch (err) {
            console.error("Failed to fetch chat data:", err);
            toast.error("Failed to load chat data. Please try again later.");
            setError("Failed to load chat data");
            onClose?.();
            return;
          }
        } else {
          setError("No users or chat ID provided");
          if (onClose) onClose();
          return;
        }
      } catch (err) {
        console.error("Error in chat initialization:", err);
        toast.error("An unexpected error occurred. Please try again later.");
        setError("Unexpected error");

        if (onClose) {
          onClose();
        }
        return;
      } finally {
        setLoading(false);
      }
      if (!error && chatId) {
        subscribeToMessages(
          chatId,
          (message) => setMessages((prev) => [...prev, message]),
          (userId) =>
            setMessages((prev) =>
              prev.map((msg) => ({ ...msg, seenBy: [...msg.seenBy, userId] })),
            ),
          (userId) =>
            setTypingUsers((prevTypingUsers) => {
              if (!prevTypingUsers.includes(userId)) {
                return [...prevTypingUsers, userId];
              }
              return prevTypingUsers;
            }),
          (userId) =>
            setTypingUsers((prev) =>
              prev.filter((preUser) => preUser != userId),
            ),
        );
        listenToOpenChatErrors();
      }
    };

    fetchData();

    return () => {
      if (chatId) {
        unsubscribeFromMessages(chatId);
      }
    };
  }, [chatId, usersId, title]);

  const messageId = Date.now().toString();

  const handleSendMessage = async (
    message: string | File | string,
    type: string,
  ) => {
    if (error || !chatId) {
      toast.error("Can't send message: chat not properly initialized");
      return;
    }

    const newMessage: any = {
      messageId,
      senderId: userIdRef.current,
      time: new Date(),
      seenBy: [],
      status: "sent",
      content: {},
    };

    switch (type) {
      case "text":
        newMessage.content.text = message;
        break;
      case "image":
        newMessage.content.image = message;
        break;
      case "video":
        newMessage.content.video = message;
        break;
      case "document":
        newMessage.content.document = message;
        break;
      default:
        console.warn("Unknown message type:", type);
        newMessage.content.text = typeof message === "string" ? message : "";
        break;
    }

    setMessages((prevMsgs) => [...prevMsgs, newMessage]);

    handleMessageStatus();
    sendMessage(chatId, newMessage, () => {
      messageDelivered.current = true;
      setMessages((prevMsgs) =>
        prevMsgs.map((msg) =>
          msg.messageId === messageId ? { ...msg, status: "delivered" } : msg,
        ),
      );
    });
  };

  const handleTypingMessage = (isTyping: boolean) => {
    if (!chatId || error) return;
    switch (isTyping) {
      case true:
        typing(chatId);
        break;
      case false:
        stopTyping(chatId);
        break;
    }
  };
  if (error) return <ChatError error={error} />;

  return (
    <>
      {chatType.current === "group" && (
        <GroupChatSetting users={users} setUsers={setUsers} chatId={chatId} />
      )}
      <div className={`${className} flex flex-col flex-1 overflow-y-hidden`}>
        <div className="h-full overflow-y-auto bg-gray-100">
          <ChatMessages users={users} messages={messages} loading={loading} />
          {typingUsers.map((typingUserId) => (
            <UserTypingIndicator
              key={typingUserId}
              userImage={
                users.find((user) => user.userId == typingUserId)
                  ?.profilePicture
              }
            />
          ))}
          <div ref={messagesEndRef} />
        </div>
        <ChatInput
          chatId={chatId}
          onSendMessage={handleSendMessage}
          onTypingMessage={handleTypingMessage}
          className=""
        />
      </div>
    </>
  );
}

export default memo(ChatContent);
