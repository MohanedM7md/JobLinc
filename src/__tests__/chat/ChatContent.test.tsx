import { render, screen, fireEvent, act } from "@testing-library/react";
import { vi, describe, it, expect, beforeEach } from "vitest";
import ChatContent from "@chatComponent/ChatContent";
import * as ChatSocket from "@services/api/ChatSocket";
import * as chatServices from "@services/api/chatServices";
import { ChatIdProvider } from "@context/ChatIdProvider";
import { NetworkUserIdProvider } from "@context/NetworkUserIdProvider";

// Mock the services
vi.mock("@services/api/ChatSocket", () => ({
  subscribeToMessages: vi.fn(),
  unsubscribeFromMessages: vi.fn(),
  sendMessage: vi.fn(),
  typing: vi.fn(),
  stopTyping: vi.fn(),
  listenToOpenChatErrors: vi.fn(),
}));

vi.mock("@services/api/chatServices", () => ({
  fetchChatData: vi.fn(),
  createChat: vi.fn(),
}));

// Mock child components
vi.mock("@chatComponent/ChatMessages", () => ({
  default: () => <div data-testid="chat-messages" />,
}));

vi.mock("@chatComponent/ChatInput", () => ({
  default: ({ onSendMessage, onTypingMessage }) => (
    <div data-testid="chat-input">
      <button
        onClick={() => onSendMessage("test message", "text")}
        data-testid="send-button"
      >
        Send
      </button>
      <button
        onClick={() => onTypingMessage(true)}
        data-testid="start-typing-button"
      >
        Start Typing
      </button>
      <button
        onClick={() => onTypingMessage(false)}
        data-testid="stop-typing-button"
      >
        Stop Typing
      </button>
    </div>
  ),
}));

vi.mock("@chatComponent/GroupSetting/GroupChatSetting", () => ({
  default: () => <div data-testid="group-chat-settings" />,
}));

vi.mock("@chatComponent/UserTyping", () => ({
  default: () => <div data-testid="user-typing-indicator" />,
}));
const mockScrollIntoView = vi.fn();
window.HTMLElement.prototype.scrollIntoView = mockScrollIntoView;
const mockChatData = {
  participants: [
    { userId: "1", profilePicture: "pic1.jpg" },
    { userId: "2", profilePicture: "pic2.jpg" },
  ],
  messages: [],
  chatId: "chat123",
  chatType: "private",
};

describe("ChatContent", () => {
  beforeEach(() => {
    vi.clearAllMocks();
    localStorage.setItem("userId", "1");
    vi.mocked(chatServices.fetchChatData).mockResolvedValue(mockChatData);
    vi.mocked(chatServices.createChat).mockResolvedValue(mockChatData);
    mockScrollIntoView.mockClear();
  });

  const renderWithProviders = (chatId = null, usersId = []) => {
    return render(
      <NetworkUserIdProvider ids={usersId}>
        <ChatIdProvider id={chatId!}>
          <ChatContent />
        </ChatIdProvider>
      </NetworkUserIdProvider>,
    );
  };

  it("initializes with existing chat", async () => {
    await act(async () => {
      renderWithProviders("existingChatId");
    });

    expect(chatServices.fetchChatData).toHaveBeenCalledWith("existingChatId");
    expect(ChatSocket.subscribeToMessages).toHaveBeenCalled();
    screen.debug();
    expect(screen.getByTestId("chat-messages")).toBeInTheDocument();
    expect(screen.getByTestId("chat-input")).toBeInTheDocument();
  });

  it("creates new chat when usersId is provided without chatId", async () => {
    // Mock successful chat creation
    (chatServices.createChat as jest.Mock).mockResolvedValue({
      data: {
        chatId: "new-chat-id",
        participants: [],
        messages: [],
        chatType: "private",
      },
    });

    await act(async () => {
      renderWithProviders(null, ["2", "3"]);
    });

    expect(chatServices.createChat).toHaveBeenCalledWith(["2", "3"], "");
    expect(ChatSocket.subscribeToMessages).toHaveBeenCalled();
  });

  it("handles typing indicators", async () => {
    await act(async () => {
      renderWithProviders("chatId");
    });

    const startTypingButton = screen.getByTestId("start-typing-button");
    const stopTypingButton = screen.getByTestId("stop-typing-button");

    fireEvent.click(startTypingButton);
    expect(ChatSocket.typing).toHaveBeenCalledWith("chatId");

    fireEvent.click(stopTypingButton);
    expect(ChatSocket.stopTyping).toHaveBeenCalledWith("chatId");
  });

  it("handles message sending", async () => {
    await act(async () => {
      renderWithProviders("chatId");
    });

    const sendButton = screen.getByTestId("send-button");

    await act(async () => {
      fireEvent.click(sendButton);
    });

    expect(ChatSocket.sendMessage).toHaveBeenCalled();
  });

  it("shows group chat settings for group chats", async () => {
    const groupChatData = { ...mockChatData, chatType: "group" };
    vi.mocked(chatServices.fetchChatData).mockResolvedValue(groupChatData);

    await act(async () => {
      renderWithProviders("groupChatId");
    });

    expect(screen.getByTestId("group-chat-settings")).toBeInTheDocument();
  });

  it("unsubscribes from messages on unmount", async () => {
    const { unmount } = await act(async () => {
      return renderWithProviders("chatId");
    });

    unmount();
    expect(ChatSocket.unsubscribeFromMessages).toHaveBeenCalledWith("chatId");
  });

  it("handles message delivery status", async () => {
    vi.useFakeTimers();
    await act(async () => {
      renderWithProviders("chatId");
    });

    const sendButton = screen.getByTestId("send-button");

    await act(async () => {
      fireEvent.click(sendButton);
    });

    // Simulate message delivery timeout
    await act(async () => {
      vi.advanceTimersByTime(1000);
    });

    // Check if message status was updated
    // You might need to modify this based on how you want to verify the status
    expect(ChatSocket.sendMessage).toHaveBeenCalled();

    vi.useRealTimers();
  });
});
