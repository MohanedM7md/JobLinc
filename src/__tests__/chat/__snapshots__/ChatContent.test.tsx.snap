// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatContent > matches snapshot 1`] = `
<div>
  <div
    class="undefined flex flex-col flex-1 overflow-y-hidden"
  >
    <div
      class="h-8/12 overflow-y-auto"
    >
      <div
        class="flex-1 space-y-2 bg-gray-100 flex flex-col pt-2 h-auto"
      >
        <div
          class="text-gray-500 text-center mt-4"
        >
          No messages found
        </div>
      </div>
    </div>
    <div
      class="flex flex-col items-center p-3 
     bg-charcoalWhite dark:bg-warmBlack border-t border-gray-200 relative"
    >
      <textarea
        class="flex-1 w-full bg-warmWhite dark:bg-darkGray text-charcoalBlack
         dark:text-charcoalWhite p-2 rounded-lg resize-none h-10 focus:outline-none
         shadow-inner transition-all ring-1 ring-gray-400"
        name="msg"
        placeholder="Write a message..."
      />
      <div
        class="flex justify-between w-full pt-4"
      >
        <div>
          <button
            class="text-mutedSilver dark:text-charcoalWhite hover:text-charcoalBlack 
            dark:hover:text-warmWhite p-2"
          >
            <svg
              class="lucide lucide-paperclip"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.234 20.252 21 12.3"
              />
              <path
                d="m16 6-8.414 8.586a2 2 0 0 0 0 2.828 2 2 0 0 0 2.828 0l8.414-8.586a4 4 0 0 0 0-5.656 4 4 0 0 0-5.656 0l-8.415 8.585a6 6 0 1 0 8.486 8.486"
              />
            </svg>
          </button>
          <button
            class="text-mutedSilver dark:text-charcoalWhite hover:text-charcoalBlack 
            dark:hover:text-warmWhite p-2"
          >
            <svg
              class="lucide lucide-image"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="3"
              />
              <circle
                cx="9"
                cy="9"
                r="2"
              />
              <path
                d="m21 15-3.086-3.086a2 2 0 0 0-2.828 0L6 21"
              />
            </svg>
          </button>
          <button
            class="text-mutedSilver dark:text-charcoalWhite hover:text-charcoalBlack 
            dark:hover:text-warmWhite p-2"
          >
            <svg
              class="lucide lucide-smile"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <path
                d="M8 14s1.5 2 4 2 4-2 4-2"
              />
              <line
                x1="9"
                x2="9.01"
                y1="9"
                y2="9"
              />
              <line
                x1="15"
                x2="15.01"
                y1="9"
                y2="9"
              />
            </svg>
          </button>
        </div>
        <button
          class="ml-2 p-1 rounded-l-full font-bold rounded-r-full transition-colors
            bg-mutedSilver text-charcoalBlack cursor-not-allowed w-15 md:w-20"
          disabled=""
          id="send-msg-btn"
        >
          Send
        </button>
      </div>
    </div>
  </div>
</div>
`;
